### PR TITLE
openPMD: Particle Filter (Container)

### LINKED ISSUE #1004: Particle filters in OpenPMD
The AMReX writer functions for the plotfile format take a particle filter, i.e., a lambda function that is evaluated on each particle and returns 0 (do not dump this particle) or 1 (dump this particle).

Class `ParticleDiag` holds the the general properties of these filters (provided by the user), and the filters are constructed in `FlushFormatPlotfile::WriteParticles`, and a lambda function containing all these filters is passed to `amrex::ParticleContainer::WritePlotFile`.

It would be useful to have this feature working for OpenPMD output.

### PR DESCRIPTION
**Pull request by ax3l**
A fresh implementation of particle filters for openPMD diagnostics using the generalized `PhysicalParticleContainer` approach that we also use in plotfiles :tada: 
Close #1004

Run with MPI on CPU and analyzed with openPMD-viewer :heavy_check_mark: 

Added a CI test that confirms filter and coarsening are working, since we had no coverage so far.

Screenshot from 2021-01-15 16-07-37

### CODE REVIEW

File: Source/Diagnostics/WarpXOpenPMD.cpp [hunk 1]
```diff
@@ -333,15 +334,47 @@ WarpXOpenPMDPlot::WriteOpenPMDParticles (const amrex::Vector<ParticleDiag>& part
        // integer attribs, and it is automatically dumped as particle record
        // when ionization is on.
        int_flags.resize(1, 1);
+       tmp.AddIntComp(false);
     }
 
-    // Convert momentum to SI
-    pc->ConvertUnits(ConvertDirection::WarpX_to_SI);
+#ifdef WARPX_QED
+      if( pc->has_breit_wheeler() ) {
```
Comment (ax3l) [inline Source/Diagnostics/WarpXOpenPMD.cpp:342]: @lucafedeli88 @NeilZaim I think I forgot some QED attributes here in the past for openPMD output. This PR should also add support for those now.

**Review by atmyers — APPROVED** [review]

**Comment by ax3l** [comment]
Although I had no crash locally, I debugged now in Debug mode and can see that both in plotfiles and openPMD output the following assert raises as soon as a `<diag>.<species>.plot_filter_function` is defined:
```
  Writing plotfile LaserIonAcc2d_plt00000
0::Assertion `index < AMREX_SPACEDIM' failed, file "/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_Particle.H", line 227 !!!
SIGABRT
```
```
=== If no file names and line numbers are shown below, one can run
            addr2line -Cpfie my_exefile my_line_address
    to convert `my_line_address` (e.g., 0x4a6b) into file name and line number.
    Or one can use amrex/Tools/Backtrace/parse_bt.py.

=== Please note that the line number reported by addr2line may not be accurate.
    One can use
            readelf -wl my_exefile | grep my_line_address'
    to find out the offset for that line.

 0: ./warpx(+0x5b06f6) [0x5626d67456f6]
    amrex::BLBackTrace::print_backtrace_info(_IO_FILE*)
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX_BLBackTrace.cpp:175:29

 1: ./warpx(+0x5b0111) [0x5626d6745111]
    amrex::BLBackTrace::handler(int)
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX_BLBackTrace.cpp:85:8

 2: ./warpx(+0x496688) [0x5626d662b688]
    amrex::ParallelDescriptor::Abort(int, bool)
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX_ParallelDescriptor.cpp:206:22

 3: ./warpx(+0x44be29) [0x5626d65e0e29]
    amrex::Assert_host(char const*, char const*, int, char const*)
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX.cpp:310:9

 4: ./warpx(+0x14df7d) [0x5626d62e2f7d]
    amrex::Assert(char const*, char const*, int, char const*) inlined at /home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_Particle.H:227:9 in amrex::Particle<4, 0>::pos(int) const &
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX.H:192:5
amrex::Particle<4, 0>::pos(int) const &
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_Particle.H:227:9

 5: ./warpx(+0x204834) [0x5626d6399834]
    FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const::{lambda(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&)#1}::operator()(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&) const
/home/axel/src/warpx/Source/Particles/Filter/FilterFunctors.H:110:37

 6: ./warpx(+0x208bb9) [0x5626d639dbb9]
    amrex::filterParticles<amrex::ParticleTile<0, 0, 4, 0, std::allocator>, amrex::ParticleTile<0, 0, 4, 0, std::allocator>, FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const::{lambda(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&)#1}, int, int>(amrex::ParticleTile<0, 0, 4, 0, std::allocator>&, FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const::{lambda(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&)#1} const&, FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const::{lambda(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&)#1}&&, int, FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const::{lambda(amrex::ConstParticleTileData<0, 0, 4, 0> const&, int, amrex::RandomEngine const&)#1}&&, int)::{lambda(int, amrex::RandomEngine const&)#1}::operator()(int, amrex::RandomEngine const&) const
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_ParticleUtil.H:45:26

 7: ./warpx(+0x20919f) [0x5626d639e19f]
    ParallelForRNG<int, amrex::filterParticles(DstTile&, const SrcTile&, Pred&&, Index, Index, N) [with DstTile = amrex::ParticleTile<0, 0, 4, 0, std::allocator>; SrcTile = amrex::ParticleTile<0, 0, 4, 0, std::allocator>; Pred = FlushFormatPlotfile::WriteParticles(const string&, const amrex::Vector<ParticleDiag>&) const::<lambda(const SrcData&, int, const amrex::RandomEngine&)>; Index = int; N = int]::<lambda(int, const amrex::RandomEngine&)> >
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Base/AMReX_GpuLaunchFunctsC.H:585:5

 8: ./warpx(+0x208da1) [0x5626d639dda1]
    filterParticles<amrex::ParticleTile<0, 0, 4, 0, std::allocator>, amrex::ParticleTile<0, 0, 4, 0, std::allocator>, FlushFormatPlotfile::WriteParticles(const string&, const amrex::Vector<ParticleDiag>&) const::<lambda(const SrcData&, int, const amrex::RandomEngine&)>, int, int>
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_ParticleTransformation.H:400:27

 9: ./warpx(+0x20839c) [0x5626d639d39c]
    addParticles<FlushFormatPlotfile::WriteParticles(const string&, const amrex::Vector<ParticleDiag>&) const::<lambda(const SrcData&, int, const amrex::RandomEngine&)> >
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_ParticleContainerI.H:1011:48

10: ./warpx(+0x207e6d) [0x5626d639ce6d]
    copyParticles<FlushFormatPlotfile::WriteParticles(const string&, const amrex::Vector<ParticleDiag>&) const::<lambda(const SrcData&, int, const amrex::RandomEngine&)> >
/home/axel/src/warpx/build/_deps/fetchedamrex-src/Src/Particle/AMReX_ParticleContainerI.H:981:5

11: ./warpx(+0x2050ed) [0x5626d639a0ed]
    FlushFormatPlotfile::WriteParticles(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&) const
/home/axel/src/warpx/Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp:333:29

12: ./warpx(+0x202e3b) [0x5626d6397e3b]
    FlushFormatPlotfile::WriteToFile(amrex::Vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > >, amrex::Vector<amrex::MultiFab, std::allocator<amrex::MultiFab> > const&, amrex::Vector<amrex::Geometry, std::allocator<amrex::Geometry> >&, amrex::Vector<int, std::allocator<int> >, double, amrex::Vector<ParticleDiag, std::allocator<ParticleDiag> > const&, int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, bool, bool, bool, bool) const
/home/axel/src/warpx/Source/Diagnostics/FlushFormats/FlushFormatPlotfile.cpp:49:17

13: ./warpx(+0xc96cb) [0x5626d625e6cb]
    FullDiagnostics::Flush(int)
/home/axel/src/warpx/Source/Diagnostics/FullDiagnostics.cpp:101:32

14: ./warpx(+0xbb2ff) [0x5626d62502ff]
    Diagnostics::FilterComputePackFlush(int, bool)
/home/axel/src/warpx/Source/Diagnostics/Diagnostics.cpp:319:18

15: ./warpx(+0xd9280) [0x5626d626e280]
    MultiDiagnostics::FilterComputePackFlush(int, bool)
/home/axel/src/warpx/Source/Diagnostics/MultiDiagnostics.cpp:67:23

16: ./warpx(+0x2fdd80) [0x5626d6492d80]
    WarpX::InitData()
/home/axel/src/warpx/Source/Initialization/WarpXInitData.cpp:113:13

17: ./warpx(+0x27eb5) [0x5626d61bceb5]
    main
/home/axel/src/warpx/Source/main.cpp:52:21

18: /lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0xf3) [0x7f439e1730b3]
    __libc_start_main
../csu/libc-start.c:342:3

19: ./warpx(+0x27bde) [0x5626d61bcbde]
    _start
??:0


===== TinyProfilers ======
main()
WarpX::InitData()
Diagnostics::FilterComputePackFlush()
FlushFormatPlotfile::WriteToFile()
ParticleContainer::copyParticles
ParticleContainer::addParticles
```

**Comment by ax3l** [comment]
That might explain the CI crashes we had in #1384 - we have a hidden out-of-bounds access in `copyParticles` with filters.

**Comment by atmyers** [comment]
The problem is here, in `Source/Particles/Filter/FilterFunctors.H:

```
    /**
     * \brief return 1 if the particle is selected by the parser
     * \param p one particle
     * \return whether or not the particle is selected
     */
    AMREX_GPU_HOST_DEVICE AMREX_FORCE_INLINE
    bool operator () (const SuperParticleType& p, const amrex::RandomEngine&) const noexcept
    {
        if ( !m_is_active ) return 1;
        amrex::Real const x  = p.pos(0);
        amrex::Real const y  = p.pos(1);
        amrex::Real const z  = p.pos(2);
```

The last line is out of bounds in 2D. We should be going through the `GetParticlePosition` functor instead of accessing the position directly.

**Comment by ax3l** [comment]
Just spotted the same location! That's the one :)

**Comment by guj** [comment]
there is a real property, the FlushFormatPlotFile.cpp:276 says "weight", while openPMD says "weighting",  I assume they should be consistent? 

Also, since the body to get particle property names are the same, would it be good to put it somewhere, for example, the FlushForrmat class? so the supported formats can call the same function? @ax3l

**Comment by ax3l** [comment]
Yes, in openPMD we standardized the name to be `weighting`, thus the slight difference:
https://github.com/openPMD/openPMD-standard/blob/upcoming-2.0.0/EXT_ED-PIC.md#namings-for-records-per-particle-species

**Comment by ax3l** [comment]
Agreed to refactor out the name setup for real and int names, because we otherwise tend to forget to add them.
Same for the Ascent flush format. Do you want to tackle this in a follow-up PR @guj? :)

**Pull request merged**